Saturday night coding fever

brunoguindani opened the pull request:
Just an hour's worth of light-hearted coding to rest a bit after the big push (or pull, rather 😏). This is stuff that needed to be sooner or later:
* load files now use the string returned by the objects' `get_id()` as the factory identifier, so that we don't have double the amount of strings to take care of
* removed `set_n_aux()` from the base class; for Neal8 this parameter is set via the now well-known dynamic casting trick
* changed HPP to H in all header guards. This is in anticipation of another adaptation to the Google style guide: renaming all .hpp files to .h and .cpp files to .cc. Let me know when I should do this, the best moment should be when you're not working on anything on the C++ side so that we don't get stuck with endless merging
* hierarchy `*_lpdf_grid()` functions have been moved to the base classes, and they always call the current class' `*_lpdf()` function; they are still virtual so that optimized versions can be built in the future
* uniformed "univ" -> "uni" in several proto files. I know that you don't like renaming protos, but we had some uni's and some univ's and it was getting really confusing
* moved mixings mass functions from their .hpp to their .cpp files.

I still have some of these miscellanea left, but they're best reserved for later down the line.

Lastly, if you have some of your own, feel free to tell me, this is the right PR for it!

mberaha commented:
Great, thanks!

Since we're changing namings, I wanted at some point to replace the algos', collectors' etc. names with Enums, for consistency also across platforms (I mainly have in mind pybmix), and have the Builder classes  working with protobuf Enums instead of string identifiers.

I mainly have in mindi protobuf Enums instead of C++ Enums because they have "easy" way of converting strings to int, so that we can pass strings as command line arguments and check if they match an Enum.

See
https://developers.google.com/protocol-buffers/docs/proto3#enum
https://developers.google.com/protocol-buffers/docs/cpptutorial#enums-and-nested-classes
https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#enum
https://stackoverflow.com/a/46470954/5609462

mberaha commented:
Also feel free to rename all the cpp files, I only have #41 to work on right now, but it's mainly CMake.

brunoguindani commented:
> Since we're changing namings, I wanted at some point to replace the algos', collectors' etc. names with Enums, for consistency also across platforms (I mainly have in mind pybmix), and have the Builder classes working with protobuf Enums instead of string identifiers.

Would something like this work?:
```
syntax = "proto3";

package bayesmix;

message RuntimeChoice {
  enum Algorithms {
    Neal2 = 0;
    Neal8 = 1;
  }
  enum Hierarchies {
    NNIG = 0;
    NNW = 1;
    LinRegUni = 2;
  }
  enum Mixings {
    DP = 0;
    PY = 1;
  }
  oneof choice {
    Algorithms algorithm = 1;
    Hierarchies hierarchy = 2;
    Mixings mixing = 3;
  }
}
```
Then we declare the factory Identifier to be the `RuntimeChoice` class and use one of its subclasses in each of the load files? Do we have to touch the library classes at all?

brunoguindani commented:
In the meantime, here's a recap of further changes:
* renamed files as planned
* Neal8 now also uses the new functions introduced in the last PR, or rather their overridden version
* updated include paths removing all ".." remaining
* added conjugacy flag and checks which for some reason we didn't have already
* indentation and formatting stuff

mberaha commented:
After second thoughts, I guess having proto Enums is more work than I anticipated.
I'm still planning to do so, but it might take me some time to figure out the right strategy. In the meantime feel free to merge
